File: .ai/scripts/verify_anim_compat.py
"""
Blender Python 腳本：驗證每個 zombie 模型的動畫骨架是否與 mesh 骨架吻合。
用法：blender --background --python verify_anim_compat.py
"""
import bpy
import os
import sys
import json

MODELS_DIR = r"D:\GlobalGanLan\public\models"
ANIM_TYPES = ["idle", "attack", "hurt", "dying", "run"]

def clear_scene():
    bpy.ops.wm.read_factory_settings(use_empty=True)

def get_armature_bones(glb_path):
    """載入 GLB 並提取骨架的骨頭名稱列表"""
    clear_scene()
    try:
        bpy.ops.import_scene.gltf(filepath=glb_path)
    except Exception as e:
        return None, str(e)
    
    # 找到 Armature
    armatures = [obj for obj in bpy.data.objects if obj.type == 'ARMATURE']
    if not armatures:
        return None, "NO_ARMATURE"
    
    armature = armatures[0]
    bone_names = sorted([b.name for b in armature.data.bones])
    return bone_names, None

def get_animation_info(glb_path):
    """載入 GLB 並提取動畫資訊（骨頭名 + 動畫名 + 幀數）"""
    clear_scene()
    try:
        bpy.ops.import_scene.gltf(filepath=glb_path)
    except Exception as e:
        return None, None, 0, str(e)
    
    armatures = [obj for obj in bpy.data.objects if obj.type == 'ARMATURE']
    if not armatures:
        return None, None, 0, "NO_ARMATURE"
    
    armature = armatures[0]
    bone_names = sorted([b.name for b in armature.data.bones])
    
    # 取得動畫
    actions = list(bpy.data.actions)
    anim_name = actions[0].name if actions else "NO_ACTION"
    frame_count = 0
    if actions:
        action = actions[0]
        frame_range = action.frame_range
        frame_count = int(frame_range[1] - frame_range[0])
    
    return bone_names, anim_name, frame_count, None

results = {}

for i in range(1, 31):
    zombie_id = f"zombie_{i}"
    model_dir = os.path.join(MODELS_DIR, zombie_id)
    
    if not os.path.isdir(model_dir):
        results[zombie_id] = {"error": "DIR_NOT_FOUND"}
        continue
    
    mesh_path = os.path.join(model_dir, f"{zombie_id}.glb")
    if not os.path.exists(mesh_path):
        results[zombie_id] = {"error": "MESH_NOT_FOUND"}
        continue
    
    # 取得 mesh 骨架
    mesh_bones, err = get_armature_bones(mesh_path)
    if err:
        results[zombie_id] = {"error": f"MESH_LOAD_ERROR: {err}"}
        continue
    
    entry = {
        "mesh_bone_count": len(mesh_bones),
        "animations": {}
    }
    
    for anim_type in ANIM_TYPES:
        anim_path = os.path.join(model_dir, f"{zombie_id}_{anim_type}.glb")
        if not os.path.exists(anim_path):
            entry["animations"][anim_type] = {"status": "FILE_MISSING"}
            continue
        
        anim_bones, anim_name, frame_count, err = get_animation_info(anim_path)
        if err:
            entry["animations"][anim_type] = {"status": f"LOAD_ERROR: {err}"}
            continue
        
        # 比對骨架
        if anim_bones == mesh_bones:
            entry["animations"][anim_type] = {
                "status": "OK",
                "anim_name": anim_name,
                "frames": frame_count,
                "bone_count": len(anim_bones)
            }
        else:
            # 找出差異
            mesh_set = set(mesh_bones)
            anim_set = set(anim_bones)
            only_in_mesh = sorted(mesh_set - anim_set)
            only_in_anim = sorted(anim_set - mesh_set)
            common = sorted(mesh_set & anim_set)
            
            entry["animations"][anim_type] = {
                "status": "MISMATCH",
                "anim_name": anim_name,
                "frames": frame_count,
                "mesh_bone_count": len(mesh_bones),
                "anim_bone_count": len(anim_bones),
                "common_bones": len(common),
                "only_in_mesh": only_in_mesh[:10],  # 只列前10個
                "only_in_anim": only_in_anim[:10],
                "only_in_mesh_count": len(only_in_mesh),
                "only_in_anim_count": len(only_in_anim)
            }
    
    results[zombie_id] = entry

# 輸出結果
output_path = os.path.join(MODELS_DIR, "..", "anim_compat_report.json")
with open(output_path, "w", encoding="utf-8") as f:
    json.dump(results, f, indent=2, ensure_ascii=False)

# 摘要
print("\n" + "="*60)
print("動畫吻合度檢查報告")
print("="*60)

issues = []
for zombie_id in sorted(results.keys(), key=lambda x: int(x.split("_")[1])):
    entry = results[zombie_id]
    if "error" in entry:
        print(f"  {zombie_id}: ERROR - {entry['error']}")
        issues.append(zombie_id)
        continue
    
    anims = entry.get("animations", {})
    zombie_ok = True
    for anim_type, info in sorted(anims.items()):
        status = info.get("status", "UNKNOWN")
        if status == "OK":
            pass  # good
        elif status == "FILE_MISSING":
            pass  # acceptable for run.glb on zombie_16~30
        elif status == "MISMATCH":
            zombie_ok = False
            mesh_bc = info.get("mesh_bone_count", 0)
            anim_bc = info.get("anim_bone_count", 0)
            common = info.get("common_bones", 0)
            only_mesh = info.get("only_in_mesh_count", 0)
            only_anim = info.get("only_in_anim_count", 0)
            print(f"  {zombie_id}/{anim_type}: MISMATCH - mesh:{mesh_bc} anim:{anim_bc} common:{common} only_mesh:{only_mesh} only_anim:{only_anim}")
            if info.get("only_in_mesh"):
                print(f"    only_in_mesh: {info['only_in_mesh']}")
            if info.get("only_in_anim"):
                print(f"    only_in_anim: {info['only_in_anim']}")
        else:
            zombie_ok = False
            print(f"  {zombie_id}/{anim_type}: {status}")
    
    if not zombie_ok:
        issues.append(zombie_id)

print("\n" + "="*60)
if issues:
    print(f"有問題的模型 ({len(issues)}): {', '.join(issues)}")
else:
    print("全部 30 個模型的動畫骨架均吻合 ✓")
print("="*60)
print(f"\n詳細報告: {output_path}")

File: .ai/scripts/verify_anim_compat.mjs
/**
 * 驗證每個 zombie 模型的動畫骨架是否與 mesh 骨架吻合
 * 直接解析 GLB 二進制格式，不需要 Blender
 * 用法: node verify_anim_compat.mjs
 */
import fs from 'fs';
import path from 'path';

const MODELS_DIR = 'D:\\GlobalGanLan\\public\\models';
const ANIM_TYPES = ['idle', 'attack', 'hurt', 'dying', 'run'];

/**
 * 解析 GLB 檔案，提取 glTF JSON 部分
 */
function parseGlb(filePath) {
  const buf = fs.readFileSync(filePath);
  // GLB header: magic(4) + version(4) + length(4) = 12 bytes
  const magic = buf.readUInt32LE(0);
  if (magic !== 0x46546C67) throw new Error('Not a GLB file');
  
  // Chunk 0: JSON
  const chunk0Length = buf.readUInt32LE(12);
  const chunk0Type = buf.readUInt32LE(16);
  if (chunk0Type !== 0x4E4F534A) throw new Error('First chunk is not JSON');
  
  const jsonStr = buf.toString('utf-8', 20, 20 + chunk0Length);
  return JSON.parse(jsonStr);
}

/**
 * 從 glTF JSON 提取骨架關節名稱
 */
function extractSkeletonInfo(gltf) {
  const nodes = gltf.nodes || [];
  const skins = gltf.skins || [];
  const animations = gltf.animations || [];
  
  // 從 skins 取得 joints
  const jointIndices = new Set();
  for (const skin of skins) {
    if (skin.joints) {
      for (const j of skin.joints) jointIndices.add(j);
    }
  }
  
  // 骨骼名稱
  const jointNames = [];
  for (const idx of jointIndices) {
    if (nodes[idx]) {
      jointNames.push(nodes[idx].name || `node_${idx}`);
    }
  }
  jointNames.sort();
  
  // 動畫資訊
  const animInfo = animations.map(a => ({
    name: a.name || 'unnamed',
    channels: a.channels?.length || 0,
    samplers: a.samplers?.length || 0
  }));
  
  // 從動畫 channels 提取被動畫的節點
  const animatedNodeIndices = new Set();
  for (const anim of animations) {
    for (const ch of (anim.channels || [])) {
      if (ch.target?.node !== undefined) {
        animatedNodeIndices.add(ch.target.node);
      }
    }
  }
  const animatedNodeNames = [];
  for (const idx of animatedNodeIndices) {
    if (nodes[idx]) {
      animatedNodeNames.push(nodes[idx].name || `node_${idx}`);
    }
  }
  animatedNodeNames.sort();
  
  // meshes
  const meshNodeNames = [];
  for (const node of nodes) {
    if (node.mesh !== undefined) {
      meshNodeNames.push(node.name || 'unnamed_mesh');
    }
  }
  
  return {
    jointNames,
    jointCount: jointNames.length,
    animatedNodeNames,
    animInfo,
    meshNodeNames,
    totalNodes: nodes.length,
    skinCount: skins.length
  };
}

// ===== 主程式 =====
const results = {};
let issueCount = 0;

for (let i = 1; i <= 30; i++) {
  const zombieId = `zombie_${i}`;
  const modelDir = path.join(MODELS_DIR, zombieId);
  
  if (!fs.existsSync(modelDir)) {
    results[zombieId] = { error: 'DIR_NOT_FOUND' };
    continue;
  }
  
  const meshPath = path.join(modelDir, `${zombieId}.glb`);
  if (!fs.existsSync(meshPath)) {
    results[zombieId] = { error: 'MESH_NOT_FOUND' };
    continue;
  }
  
  let meshInfo;
  try {
    const meshGltf = parseGlb(meshPath);
    meshInfo = extractSkeletonInfo(meshGltf);
  } catch (e) {
    results[zombieId] = { error: `MESH_PARSE_ERROR: ${e.message}` };
    continue;
  }
  
  const entry = {
    mesh_joints: meshInfo.jointCount,
    mesh_meshes: meshInfo.meshNodeNames,
    animations: {}
  };
  
  for (const animType of ANIM_TYPES) {
    const animPath = path.join(modelDir, `${zombieId}_${animType}.glb`);
    if (!fs.existsSync(animPath)) {
      entry.animations[animType] = { status: 'FILE_MISSING' };
      continue;
    }
    
    let animInfo;
    try {
      const animGltf = parseGlb(animPath);
      animInfo = extractSkeletonInfo(animGltf);
    } catch (e) {
      entry.animations[animType] = { status: `PARSE_ERROR: ${e.message}` };
      continue;
    }
    
    // 比對骨架
    const meshJointSet = new Set(meshInfo.jointNames);
    const animJointSet = new Set(animInfo.jointNames);
    
    // 如果動畫 GLB 沒有 skin（純動畫，無幾何體），用 animatedNodeNames 比對
    let comparisonTarget;
    let comparisonLabel;
    if (animInfo.jointCount === 0 && animInfo.animatedNodeNames.length > 0) {
      // 純動畫檔 — 比對被動畫的節點是否都存在於 mesh 的 joints 中
      comparisonTarget = animInfo.animatedNodeNames;
      comparisonLabel = 'animated_nodes';
    } else {
      comparisonTarget = animInfo.jointNames;
      comparisonLabel = 'joints';
    }
    
    const compSet = new Set(comparisonTarget);
    const onlyInMesh = meshInfo.jointNames.filter(n => !compSet.has(n));
    const onlyInAnim = comparisonTarget.filter(n => !meshJointSet.has(n));
    const common = meshInfo.jointNames.filter(n => compSet.has(n));
    
    if (onlyInMesh.length === 0 && onlyInAnim.length === 0) {
      entry.animations[animType] = {
        status: 'PERFECT_MATCH',
        anim_names: animInfo.animInfo.map(a => a.name),
        comparison: comparisonLabel,
        joint_count: meshInfo.jointCount,
        has_mesh: animInfo.meshNodeNames.length > 0
      };
    } else if (onlyInAnim.length === 0 && onlyInMesh.length > 0) {
      // 動畫骨架是 mesh 骨架的子集 — 通常 OK（動畫沒有動到所有骨頭）
      entry.animations[animType] = {
        status: 'SUBSET_OK',
        anim_names: animInfo.animInfo.map(a => a.name),
        comparison: comparisonLabel,
        mesh_joints: meshInfo.jointCount,
        anim_count: comparisonTarget.length,
        common: common.length,
        unused_in_anim: onlyInMesh.length
      };
    } else {
      // 有不匹配的骨頭
      entry.animations[animType] = {
        status: 'MISMATCH',
        anim_names: animInfo.animInfo.map(a => a.name),
        comparison: comparisonLabel,
        mesh_joints: meshInfo.jointCount,
        anim_count: comparisonTarget.length,
        common: common.length,
        only_in_mesh: onlyInMesh.slice(0, 15),
        only_in_anim: onlyInAnim.slice(0, 15),
        only_in_mesh_count: onlyInMesh.length,
        only_in_anim_count: onlyInAnim.length
      };
    }
  }
  
  results[zombieId] = entry;
}

// ===== 輸出報告 =====
const reportPath = path.join(MODELS_DIR, '..', 'anim_compat_report.json');
fs.writeFileSync(reportPath, JSON.stringify(results, null, 2));

console.log('\n' + '='.repeat(70));
console.log('動畫骨架吻合度檢查報告');
console.log('='.repeat(70));

for (let i = 1; i <= 30; i++) {
  const zombieId = `zombie_${i}`;
  const entry = results[zombieId];
  
  if (entry.error) {
    console.log(`  ❌ ${zombieId}: ${entry.error}`);
    issueCount++;
    continue;
  }
  
  const anims = entry.animations || {};
  let hasIssue = false;
  const statusParts = [];
  
  for (const animType of ANIM_TYPES) {
    const info = anims[animType];
    if (!info) continue;
    
    switch (info.status) {
      case 'PERFECT_MATCH':
        statusParts.push(`${animType}:✅`);
        break;
      case 'SUBSET_OK':
        statusParts.push(`${animType}:⚠️(${info.common}/${info.mesh_joints})`);
        break;
      case 'FILE_MISSING':
        statusParts.push(`${animType}:⬜`);
        break;
      case 'MISMATCH':
        hasIssue = true;
        statusParts.push(`${animType}:❌(mesh:${info.only_in_mesh_count} anim:${info.only_in_anim_count})`);
        break;
      default:
        hasIssue = true;
        statusParts.push(`${animType}:❓${info.status}`);
    }
  }
  
  const icon = hasIssue ? '❌' : '✅';
  console.log(`  ${icon} ${zombieId} [${entry.mesh_joints} joints, mesh: ${entry.mesh_meshes?.join(',')}]`);
  console.log(`     ${statusParts.join(' | ')}`);
  
  // 列出不匹配的詳情
  if (hasIssue) {
    issueCount++;
    for (const animType of ANIM_TYPES) {
      const info = anims[animType];
      if (info?.status === 'MISMATCH') {
        console.log(`     ── ${animType} MISMATCH 詳情:`);
        if (info.only_in_mesh?.length) {
          console.log(`        僅在 mesh: ${info.only_in_mesh.join(', ')}`);
        }
        if (info.only_in_anim?.length) {
          console.log(`        僅在 anim: ${info.only_in_anim.join(', ')}`);
        }
      }
    }
  }
}

console.log('\n' + '='.repeat(70));
console.log(`結果：${30 - issueCount}/30 模型動畫吻合 | ${issueCount} 個有問題`);
console.log('='.repeat(70));
